File: pkg/R/circulant.R
## ID: circulant.R, last updated 2022-02-26, F.Osorio

circulant <- function(x)
{ ## function returns the n-by-n circulant matrix based on 'x'
  if (!is.vector(x))
    stop("supply a vector-like 'x'")
  if (!is.numeric(x))
    stop("argument x is not a numeric vector" )
  n <- length(x)
  z <- .Fortran("circulant_mat",
                x = as.double(x),
                n = as.integer(n),
                mat = double(n * n),
                ldmat = as.integer(n),
                info = as.integer(0))
  if (z$info)
    stop(paste("circulant_mat gave error code", z$info))
  z <- matrix(z$mat, nrow = n, ncol = n)
  z
}

File: pkg/src/circulant.f
c ID: circulant.f, last updated 2022-02-26, F.Osorio

cccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
      SUBROUTINE circulant_mat(x, n, c, ldc, info)
      INTEGER          n, ldc, info
      DOUBLE PRECISION c(ldc,*), x(*)
c
c     contructs the n-by-n circular matrix based on the v vector
c
c     parameters:
c     x       (input) DOUBLE PRECISION array, dimension (n)
c             first column of the circular matrix
c     n       (input) INTEGER
c             length of vector x. n > 0.
c     c       (input/output) DOUBLE PRECISION array, dimension (ldc, n)
c             a square matrix. on exit if info = 0, is the circulant matrix.
c     ldc     (input) INTEGER
c             ldc is the leading dimension of the array c. ldc >= max(1,n).
c     info    (output) INTEGER
c             = 0: successful exit
c             < 0: if info = -i, the i-th argument had an illegal value
c
c     .. local scalars ..
      INTEGER          j
c
c     test the input parameters
c
      info = 0
      if (n .LT. 0) then
        info = -2
      else if (ldc .LT. max(1, n)) then
        info = -4
      end if
      if (info .NE. 0) return
c
c     quick return if possible
c
      if ((n .EQ. 0) .OR. (ldc .EQ. 0)) return
c
c     setup circular matrix c, column by column
c
      call dcopy(n, x(1), 1, c(1,1), 1)
      do j = 2, n
        call dcopy(n - 1, c(2,j-1), 1, c(1,j), 1)
        c(n,j) = c(1,j-1)
      end do
c
      return
      END
